src/utilities/vc_ins_utilities.cpp

```diff
@@ -0,0 +1,908 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (c) 2020 - 2024 by the IBAMR developers
+// All rights reserved.
+//
+// This file is part of IBAMR.
+//
+// IBAMR is free software and is distributed under the 3-clause BSD
+// license. The full text of the license can be found in the file
+// COPYRIGHT at the top level directory of IBAMR.
+//
+// ---------------------------------------------------------------------
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+#include "ibamr/vc_ins_utilities.h"
+
+#include "ibtk/PhysicalBoundaryUtilities.h"
+#include <ibtk/HierarchyMathOps.h>
+#include <ibtk/IBTK_MPI.h>
+#include <ibtk/ibtk_utilities.h>
+
+#include <ibamr/app_namespaces.h>
+
+/////////////////////////////// NAMESPACE ////////////////////////////////////
+
+namespace IBAMR
+{
+
+namespace VcINSUtilities
+{
+void
+callSetDensityCallbackFunction(int rho_idx,
+                               Pointer<Variable<NDIM> > rho_var,
+                               Pointer<IBTK::HierarchyMathOps> hier_math_ops,
+                               const int cycle_num,
+                               const double time,
+                               const double current_time,
+                               const double new_time,
+                               void* ctx)
+{
+    // Set the density from the level set information
+    static SetFluidProperties* ptr_SetFluidProperties = static_cast<SetFluidProperties*>(ctx);
+    const std::string& num_phases = ptr_SetFluidProperties->getNumberOfPhases();
+    if (num_phases == "TWO_PHASE")
+    {
```

Comment:
Maybe create an enum for TWO_PHASE and THREE_PHASE instead of string. You can place the enum in the header of this class.